File: CS355/program2.c
/*
Simulate a pond with algae AND FISH and growth therein
John Phillip Robertson
04/11/2020
Note: the Python version of this,
without even fish logic,
took 178146.402 seconds,
nearly 50 hours.
I tried to rewrite it to .cpp,
but moat wouldn't accept it because of the
<random> library,
which was the only C++ feature I was using.
To compile:
	Go to moat (or your Linux terminal of choice)
	with this file and "gcc program2.c"
To run:
	./a.out
*/
/*Sources:
https://www.geeksforgeeks.org/generating-random-number-range-c/
https://stackoverflow.com/a/3025106/9295513
https://www.tutorialspoint.com/how-do-i-generate-random-floats-in-cplusplus
*/

#include <stdlib.h>
#include <stdio.h>

#define true 1
#define false 0

typedef struct {
	int alga;
	int fish[2];
} fishalg;

fishalg lake[100][100][100];
//https://stackoverflow.com/a/3025106/9295513
#define alg_die 0.1
#define alg_reproduce (0.2 + alg_die)
#define alg_survive (0.1 + alg_reproduce)
int fish_count = 1000000;  //Alternate 1000000

void populate() {
    int fi, fj, fk;
    for (fi = 0; fi < 100; fi++) {
    	for (fj = 0; fj < 100; fj++) {
    		for (fk = 0; fk < 100; fk++) {
    			lake[fi][fj][fk] = (fishalg) {2, {0, 0}};
    		}
    	}
    }
}

int fish_and_alga_stuff(fishalg* l) {
    //https://www.tutorialspoint.com/how-do-i-generate-random-floats-in-cplusplus
    //Once again, this would otherwise be a library call.
    float chance = (((float)rand()/(float)RAND_MAX) * 1);

	//Fish logic
	l->fish[1] -= 1;
	if (l->fish[0] != 0 && l->fish[1]/l->fish[0] < 5) { //feeding logic
		if (l->alga >= 1 && chance > 0.5) {
			l->alga -= 1;
			l->fish[1] += 3;
			if (l->fish[0] != 0 && l->fish[1]/l->fish[0] > 5.0) { //satiety logic
				l->fish[1] = 5 * l->fish[0];
			}
		}
	}
    if (l->alga > 0) {
	//Now it's the alga's turn just because here is where a return can happen.
	    int alga = l->alga;
    	if (chance < alg_die) {
    		l->alga -= 1;
    		if (alga >= 5) {
    			l->alga = 4;
    		}
    	}
    	else if (alg_die < chance && chance < alg_reproduce) {
    		l->alga += 1;
    	}
    	else if (alg_survive < chance) {
    		l->alga -= 1;
    		return true;  // Our flag to know if the algae should migrate...
    	}
    	return false; // ...or not.
    }
    return false;
}

int valid_migration(int i, int j, int k) {
    if (i < 100 && i >= 0 && j < 100 && j >= 0 && k < 100 && k >= 0) {
        return true;
    }
	return false;
}

void cycle(int generations) {
    unsigned int count[] = {0, 0, 0, 0, 0};
	int step, i, j, k, f;
    for (i = 0; i < fish_count; i++) {
        //https://www.geeksforgeeks.org/generating-random-number-range-c/
        //Math formulas aren't intellectual property,
        //and if you could simply download Clang++ on moat,
        //this would just be a call to a library someone else wrote.
    	int* cache = lake[(rand() % (99 - 0 + 1)) + 0][(rand() % (99 - 0 + 1)) + 0][(rand() % (99 - 0 + 1)) + 0].fish;
    	cache[0] += 1;
    	cache[1] += 2;
    }
	for (step = 0; step < generations; step++) {
	    count[0] = 0;
	    count[1] = 0;
	    count[2] = 0;
	    count[3] = 0;
	    count[4] = 0;
		for (i = 0; i < 100; i++) {
			for (j = 0; j < 100; j++) {
				for (k = 0; k < 100; k++) {
					int result = fish_and_alga_stuff(&lake[i][j][k]);
					fishalg kalga = lake[i][j][k];
					int ip = i + (rand() % (1 - -1 + 1)) + -1;
					int jp = j + (rand() % (1 - -1 + 1)) + -1;
					int kp = k + (rand() % (1 - -1 + 1)) + -1;
					if (result) {
						if (valid_migration(ip, jp, kp)) {
							fishalg kalgap = lake[ip][jp][kp];
							kalgap.alga += 1;
							if (kalgap.alga >= 5) {
								kalgap.alga = 4;
							}
						}
					}

					if (kalga.alga == 0) {
						count[0] += 1;
					}
					else if (kalga.alga == 1) {
						count[1] += 1;
					}
					else if (kalga.alga == 2) {
						count[2] += 1;
					}
					else if (kalga.alga == 3) {
						count[3] += 1;
					}
					else if (kalga.alga == 4) {
						count[4] += 1;
					}

					//Begin fish counting
					int curr_fish = kalga.fish[0];
					for (f = 0; f < curr_fish; f++) {
						if (kalga.fish[0] > 0 && kalga.fish[1]/kalga.fish[0] < 0.0) {
							fish_count -= 1;
							continue;
						}
						ip = i + (rand() % (1 - -1 + 1)) + -1;
						jp = j + (rand() % (1 - -1 + 1)) + -1;
						kp = k + (rand() % (1 - -1 + 1)) + -1;
						if (valid_migration(ip, jp, kp)) {
							fishalg kfishp = lake[ip][jp][kp];
							kfishp.fish[0] += 1;
							fish_count += 1;
							kfishp.fish[1] += 2;
							kalga.fish[1] -= 2;
						}
					}
					if (kalga.fish[0] >= 2) {
						kalga.fish[0] += 1;
						fish_count += 1;
					}
				}
			}
		}
		printf("Generation %d: %d Fish, %u Algae (%u of 0, %u of 1, %u of 2, %u of 3, %u of 4)\n", step + 1, fish_count, count[1] + count[2] + count[3] + count[4], count[0], count[1], count[2], count[3], count[4]);
	}
}

int main() {
    populate();
	cycle(1000);
	return 0;
}

File: CS355/program2.py
"""
Simulate a pond with algae AND FISH and growth therein
John Phillip Robertson
04//2020
"""

from random import random, choice, randint, shuffle

# Event chances for algae activities
alg_die = 0.1
alg_reproduce = 0.2 + alg_die
alg_survive = 0.1 + alg_reproduce
# alg_migrate = None  # it is implicitly > alg_survive and < 1.0


class Algae():
    def __init__(self):
        self.alga = 2
        self.fish = []

    """
    Migration logic is contained in the Lake class instead.
    Death is simple: decrement alga.
    Reproduction is increment.
    """

    def step(self):
        chance = random()
        salga = self.alga
        if salga <= 3:
            if chance < alg_die:
                salga -= 1
                if salga >= 5:
                    salga = 4
            elif alg_die < chance < alg_reproduce:
                salga += 1
            # elif alg_reproduce < chance < alg_survive:
            #    pass
            elif alg_survive < chance:
                salga -= 1  # This will be incremented no matter what: either +1 for an adjacent cell, or -1 + 1 for the same cell
                return "move"


class Fish():
    def __init__(self):
        self.hp = 2

    def cycle(self, alga):
        self.hp -= 1
        if self.hp < 5:  # feeding logic
            if random() > .5:
                alga.alga -= 1
                self.hp += 3
                if self.hp > 5:  # satiety logic
                    self.hp = 5
        return self


class Lake():
    """
    The Lake class primarily consists of a 100^3 array of Algae plus a count of the fish.
    The Algae, more fittingly called "cells", contain a count of algae within
    as well as a list of Fish.
    """

    def __init__(self):
        self.fish_count = 100000  # Alternate 1000000
        self.lake = [[[Algae() for i in range(100)]
                      for j in range(100)] for k in range(100)]
        for i in range(self.fish_count):
            self.lake[randint(0, 99)][randint(
                0, 99)][randint(0, 99)].fish.append(Fish())
        # Above populates random Algae cells with fish.

    def valid_migration(self, i, j, k):
        try:
            return self.lake[i][j][k]
        except IndexError:
            # Kludge, but what can you do? If something returns, it will be truthy anyway.
            return False

    def cycle(self, steps):
        # I reset these after every iteration and count afresh.
        # The indices correspond precisely to which number is being counted.
        count = [0, 0, 0, 0, 0]
        for step in range(steps):
            for i in self.lake:
                for j in i:
                    for k in j:
                        result = k.step()
                        kalga = k.alga
                        if result == "move":
                            # a random surrounding cell is chosen, but only if it is valid
                            alGet = self.valid_migration(
                                self.lake.index(i) + choice((1, 0, -1)),
                                i.index(j) + choice((1, 0, -1)),
                                j.index(k) + choice((1, 0, -1))
                            )
                            if alGet:
                                alAl = alGet.alga
                                alAl += 1
                                if alAl >= 5:
                                    alAl = 4
                            else:
                                kalga += 1
                                if kalga >= 5:
                                    kalga = 4
                        # Every single bit of this is kludge but it's more efficient than counting post-facto
                        if kalga == 0:
                            count[0] += 1
                        elif kalga == 1:
                            count[1] += 1
                        elif kalga == 2:
                            count[2] += 1
                        elif kalga == 3:
                            count[3] += 1
                        elif kalga == 4:
                            count[4] += 1
                        # End counting. Begin fish handling.
                        kish = k.fish
                        for f in kish:
                            f.cycle(k)
                            if f.hp <= 0:
                                kish.remove(f)
                                del f
                                self.fish_count -= 1
                                continue
                            fiGet = self.valid_migration(
                                self.lake.index(i) + choice((1, 0, -1)),
                                i.index(j) + choice((1, 0, -1)),
                                j.index(k) + choice((1, 0, -1))
                            )
                            if fiGet:
                                kish.remove(f)
                                fiGet.fish.append(f)
                        if len(kish) >= 2:
                            shuffle(kish)
                            f1 = kish[0]
                            f2 = kish[1]
                            f1.hp -= 1
                            f2.hp -= 1
                            kish.extend([f1, f2])
                            kish.append(Fish())
                            self.fish_count += 1

            print(
                f"Generation {step + 1}: {self.fish_count} Fish, {count[1] + count[2] + count[3] + count[4]} Algae ({count[0]} of 0, {count[1]} of 1, {count[2]} of 2, {count[3]} of 3, {count[4]} of 4)"
            )


if __name__ == "__main__":
    green_goodness = Lake()
    green_goodness.cycle(1000)
